### PR TITLE
Feat/flat power forgetting curve

### PR DESCRIPTION
**Pull request by L-M-Sherlock**
Idea: https://github.com/open-spaced-repetition/fsrs4anki/issues/461#issuecomment-1787935005

Python Implementation: https://github.com/open-spaced-repetition/fsrs-optimizer/commit/8ac5d5bf763b2fb72857ffa3342e9b2b3a1d0d91

Help wanted: There are duplicated function named `power_forgetting_curve`. Is it possible to refactor them into one function?

### CODE REVIEW

**Comment by asukaminato0721** [comment]
linux:

```
---- pre_training::tests::test_loss stdout ----
thread 'pre_training::tests::test_loss' panicked at src/pre_training.rs:327:9:
assertion `left == right` failed
  left: 14.5771
 right: 14.577101
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

win:

```
---- pre_training::tests::test_loss stdout ----
thread 'pre_training::tests::test_loss' panicked at src\pre_training.rs:327:9:
assertion `left == right` failed
  left: 14.5771
 right: 14.577101
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

floating point again :(

**Review by asukaminato0721 — APPROVED** [review]

**Comment by dae** [comment]
I can't see an easy way to combine the three power_forgetting_curves, as you have one that uses raw floats, one that uses NdArray, and one that uses tensors. You could potentially just have a raw float one and then convert the tensors to floats and back again, but it might be a bit slower if it's called frequently.

File: src/inference.rs [hunk 1]
```diff
@@ -12,7 +12,9 @@ use crate::model::Model;
 use crate::training::BCELoss;
 use crate::{FSRSError, FSRSItem};
 use burn::tensor::ElementConversion;
-
+pub const DECAY: f64 = -0.5;
```
Comment (dae) [inline src/inference.rs:15]: pub(crate) would be a better choice

**Comment by asukaminato0721** [comment]
> Is it possible to refactor them into one function?

They have different types, not something like `Tensor<B, 1>` vs `Tensor<B, 2>`

instead, `&Array1` vs `Array1` vs `f32` vs `f64` vs `Tensor`, so it's hard to use `trait` to merge them.

Maybe first we need to unify the array type...

**Comment by asukaminato0721** [comment]
> > Is it possible to refactor them into one function?
> 
> They have different types, not something like `Tensor<B, 1>` vs `Tensor<B, 2>`
> 
> instead, `&Array1` vs `Array1` vs `f32` vs `f64` vs `Tensor`, so it's hard to use `trait` to merge them.
> 
> Maybe first we need to unify the array type...

Some small thought, maybe (I am not sure how to write the trait, this is a draft.)

```rs
fn power_forgetting_curve(a: T1, b: T2) -> T1
where
    T1: Add<Rhs = xxx, Ouput = T1> + Mul<Rhs = xxx, Output = T1> + Pow<Rhs = xxx, Output = T1>,
    T2: Add<Rhs = xxx, Ouput = T2> + Mul<Rhs = xxx, Output = T2> + Pow<Rhs = xxx, Output = T2>,
{
    a.some_operator(b)
}
```

**Pull request merged**